chain/epoch-manager/src/shard_assignment.rs

```diff
@@ -140,21 +82,311 @@ fn assign_with_possible_repeats<T: HasStake + Eq, I: Iterator<Item = (usize, T)>
+        &chunk_producers,
+        num_shards,
+        prev_chunk_producers_assignment,
+    );
+
+    // Find and assign new validators first.
+    let old_validators = chunk_producer_assignment.iter().flatten().collect::<HashSet<_>>();
+    let new_validators =
+        (0..num_chunk_producers).filter(|i| !old_validators.contains(i)).collect::<Vec<_>>();
+    let mut shard_set: BTreeSet<ShardSetItem> = (0..num_shards)
+        .map(|s| ShardSetItem {
+            shard_validators_num: chunk_producer_assignment[s as usize].len(),
+            shard_index: s as usize,
+        })
+        .collect();
+    let mut new_assignments = new_validators.len();
+    for validator_index in new_validators {
+        let ShardSetItem { shard_index, .. } = shard_set.pop_first().unwrap();
+        chunk_producer_assignment[shard_index].push(validator_index);
+        shard_set.insert(ShardSetItem {
+            shard_validators_num: chunk_producer_assignment[shard_index].len(),
+            shard_index,
+        });
+    }
+
+    // Reassign old validators to balance shards until the limit is reached.
+    let rng = &mut EpochInfo::shard_assignment_rng(&rng_seed);
+    let new_assignments_hard_limit = chunk_producers.len().max(shard_assignment_changes_limit);
+    loop {
+        let ShardSetItem {
+            shard_validators_num: minimal_shard_validators_num,
+            shard_index: minimal_shard,
+        } = *shard_set.first().unwrap();
+        let ShardSetItem {
+            shard_validators_num: maximal_shard_validators_num,
+            shard_index: maximal_shard,
+        } = *shard_set.last().unwrap();
+        let is_minimal_num_satisfied = minimal_shard_validators_num >= min_validators_per_shard;
+        let is_balanced = maximal_shard_validators_num - minimal_shard_validators_num <= 1;
+
+        if is_minimal_num_satisfied
+            && (is_balanced || new_assignments >= shard_assignment_changes_limit)
+        {
+            break;
+        }
+
+        assert!(
+            new_assignments <= new_assignments_hard_limit,
+            "Couldn't balance {num_shards} shards in {new_assignments_hard_limit}\
+             iterations. It means that some chunk producer was selected for \
+             new shard twice which shouldn't happen."
+        );
+        assert_ne!(
+            minimal_shard,
+            maximal_shard,
+            "Minimal shard and maximal shard are the same: {minimal_shard}. \
+            Either {} chunk producers are not enough to satisfy minimal number \
+            {min_validators_per_shard} for {num_shards} shards, or we try to \
+            balance the shard with itself.",
+            chunk_producers.len(),
+        );
+
+        let minimal_shard = shard_set.pop_first().unwrap().shard_index;
+        let maximal_shard = shard_set.pop_last().unwrap().shard_index;
+        let validator_pos = rng.gen_range(0..chunk_producer_assignment[maximal_shard].len());
```

Comment:
Perhaps here we could pick a validator to minimize the difference in stakes between the min_shard and max_shard. The security and balancing implications are not immediately clear to me though. No need to do it in this PR, I'm just sharing an idea to consider.